src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -298,11 +305,34 @@ private static byte[] createTemplate(ClassDesc ifaceDesc, String methodName, Lis
             clb.withSuperclass(CD_Object);
             clb.withFlags(ACC_FINAL | ACC_SYNTHETIC);
             clb.withInterfaceSymbols(ifaceDesc);
-            // <init>
-            clb.withMethodBody(INIT_NAME, MTD_void, 0, cob -> cob
-                    .aload(0)
-                    .invokespecial(CD_Object, INIT_NAME, MTD_void)
-                    .return_());
+            // <init>(Lookup)
+            clb.withMethodBody(INIT_NAME, MTD_void_Lookup, 0, cob -> {
+                cob.aload(0);
+                cob.invokespecial(CD_Object, INIT_NAME, MTD_void);
+
+                var failLabel = cob.newLabel();
+                // check lookupClass
+                cob.aload(1);
+                cob.invokevirtual(CD_MethodHandles_Lookup, "lookupClass", MTD_Class);
+                cob.constantInstruction(proxyDesc);
```

Comment:
This probably can’t use `LDC` of `CONSTANT_Class`, as that can’t refer to hidden classes:
```suggestion
                cob.aload(0);
                cob.invokevirtual(CD_Object, "getClass", MTD_Class);
```

--------------------------------------------------------------------------------

Having a helper constant bootstrap of the following form would allow to obtain the current class using a dynamic constant:
```java
public static final Class<?> lookupClass(
	final MethodHandles.Lookup lookup,
	final String name,
	@SuppressWarnings("rawtypes") // needed to allow `Class.class`
	final Class<? extends Class> type
) {
	requireNonNull(lookup);
	requireNonNull(type);
	if (type != Class.class) {
		throw new IllegalArgumentException();
	}
	return lookup.lookupClass();
}
```

Reply:
> that can’t refer to hidden classes

No, that is exactly the **one and only thing** that can refer to the hidden class:

> * On any attempt to resolve the entry in the run-time constant pool indicated by `this_class`, the symbolic reference is considered to be resolved to `C` and resolution always succeeds immediately.

[(Souce)](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#defineHiddenClass%28byte%5B%5D,boolean,java.lang.invoke.MethodHandles.Lookup.ClassOption...%29)

As the passed class refers to `this_class`, that use is fine.

Reply:
The **JVM** spec doesn’t mention that anywhere though.

Reply:
> The **JVM** spec doesn’t mention that anywhere though.

I don't think JVMS mentions hidden classes at all. You have to refer to the specification of Lookup.defineHiddenClass, as hidden classes were originated from VM anonymous classes added back in Java 7 with j.l.invoke package (JSR 292) as an implementation artifact